### PR TITLE
DOCSP-40816 - Upgrade

### PR DESCRIPTION
**Pull request by mongoKart**
This page includes breaking changes only for versions >= 3.9, since that's when the team began maintaining a changelog (and v3.0, which is flatly incompatible with prior versions).

# Pull Request Info

[PR Reviewing Guidelines](https://github.com/mongodb/docs-java/blob/master/REVIEWING.md)

JIRA - https://jira.mongodb.org/browse/DOCSP-40816
Staging - https://preview-mongodbmongokart.gatsbyjs.io/cpp-driver/docsp-40816-upgrade/upgrade/

## Self-Review Checklist

- [ ] Is this free of any warnings or errors in the RST? (still errors related to missing pages)
- [x] Did you run a spell-check?
- [x] Did you run a grammar-check?
- [x] Are all the links working?
- [x] Are the [facets and meta keywords](https://wiki.corp.mongodb.com/display/DE/Docs+Taxonomy) accurate?

### CODE REVIEW

File: source/upgrade.txt [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+Before you upgrade, perform the following actions:
+
+- Ensure the new {+driver-short+} version is compatible with the {+mdb-server+} versions
+  your application connects to and the C++ version your
+  application runs on. For version compatibility information, see the
+  :ref:`{+driver-short+} Compatibility <cpp-compatibility-tables>`
+  page.
+- Address any breaking changes between the driver version
+  your application is using and your planned upgrade version in the
+  :ref:`Breaking Changes <cpp-breaking-changes>` section.
```
Comment (mcmorisi) [inline source/upgrade.txt:35]:
```suggestion
  your application is using and your planned upgrade version as detailed in the
  :ref:`Breaking Changes <cpp-breaking-changes>` section.
```

File: source/upgrade.txt [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+
+Version 3.9 Breaking Changes
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+This driver version introduces the following breaking changes:
+
+- Removes support for exported targets from the CMake project build tree.
+- Drops support for macOS 10.14, macOS 10.15, and Ubuntu 14.04.
+- Requires MongoDB C Driver v1.25.0 or later.
+- Requires CMake v3.15 or later
```
Comment (mcmorisi) [inline source/upgrade.txt:94]:
```suggestion
- Requires CMake v3.15 or later.
```

**Review by mcmorisi — APPROVED** [review]
LGTM with a couple of nits!

File: source/upgrade.txt (hunk 2 above)
Comment (mongoKart) [reply source/upgrade.txt:94]: doh, good catch. i remember that this is the point i was questioning whether to use periods or not

File: source/upgrade.txt [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+
+.. _cpp-breaking-changes:
+
+Breaking Changes
+----------------
+
+.. note::
+
+   The {+driver-short+} doesn't ensure API or ABI stability across patch releases
+   within a minor version.
```
Comment (eramongodb) [inline source/upgrade.txt:52]:
```suggestion
   {+driver-short+} follows Semantic Versioning. Inadvertent API breaking changes may be reverted in a patch release within reason. The {+driver-short+} may break build system or package compatibility in any release. See: https://www.mongodb.com/docs/languages/cpp/cpp-driver/current/api-abi-versioning/api-versioning/
```

I do not believe this is strictly correct. Although we follow SemVer and _may_ use "best judgment" to reverting an inadvertent API breaking change in a patch release (to preserve stability with the major version number), we do not expect this to ever be necessary. API breaking changes _should_ always constitute an API major version number bump. Recommend referencing/linking the [API Versioning](https://www.mongodb.com/docs/languages/cpp/cpp-driver/current/api-abi-versioning/api-versioning/) page for more details.

ABI stability is not guaranteed for _any_ release so far, as we do not yet have a stable ABI release.

If the intent was to document potential breaking changes which include _build system_ or _packaging_ compatibility (which seems to be the case given examples of "breaking changes" below), recommend explicitly stating so as suggested. See [Build System Compatibility](https://www.mongodb.com/docs/languages/cpp/cpp-driver/current/api-abi-versioning/api-versioning/#build-system-compatibility) for reference: we explicitly do _not_ consider properties of the build system to be part of our API compatibility contract unless it directly affects the API we _do_ support stability for.

Poking @kevinAlbs for confirmation.

File: source/upgrade.txt [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+The breaking changes in this section are categorized by the driver version that introduced
+them. When upgrading driver versions, address all the breaking changes between the current
+and upgrade versions.
+
+.. example:: Upgrading from Version 3.0
+
+  If you're upgrading the {+driver-short+} from version 3.0 to version 3.10,
+  address all breaking changes listed for versions 3.1 to 3.10, if any.
+
+.. _version-3.10-breaking-changes:
```
Comment (eramongodb) [inline source/upgrade.txt:67]: > This page includes breaking changes only for versions >= 3.9, since that's when the team began maintaining a changelog (and v3.0, which is flatly incompatible with prior versions).

Suggest removing mentions of any versions prior to 3.9 if the intent is to avoid having to translate [prior release notes](https://github.com/mongodb/mongo-cxx-driver/releases#hd-a8ede57b) into a log of potential build-breaking changes. This jump from 3.0 to 3.9 may only confuse users due to the absence of any intermediate releases.

File: source/upgrade.txt [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+
+Version 3.10 Breaking Changes
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+This driver version introduces the following breaking changes:
+
+- Drops support for the ``find_package(libbsoncxx)`` and ``find_package(libmongocxx)`` CMake
+  commands. Call ``find_package(bsoncxx)`` and ``find_package(mongocxx)``
+  instead.
```
Comment (eramongodb) [inline source/upgrade.txt:76]:
```suggestion
- Drops support for ``find_package(libbsoncxx)`` and ``find_package(libmongocxx)``. Use ``find_package(bsoncxx)`` and ``find_package(mongocxx)``
  instead.
```
Slight wording tweaks to avoid confusing the package with the command.

**Review by eramongodb — CHANGES_REQUESTED** [review]

File: source/upgrade.txt (hunk 3 above)
Comment (kevinAlbs) [reply source/upgrade.txt:52]: Suggested wording matches my expectation. I consider an API break in a minor release a bug. That bug may be fixed in a patch release.

File: source/upgrade.txt [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+--------
+
+This page describes the changes you must make to your application
+when you upgrade to a new version of the {+driver-short+}.
+
+Before you upgrade, perform the following actions:
+
+- Ensure the new {+driver-short+} version is compatible with the {+mdb-server+} versions
+  your application connects to and the C++ version your
+  application runs on. For version compatibility information, see the
```
[the remark this answers is not in the document]
Comment (mongoKart) [reply source/upgrade.txt:30]: Incorporated suggested wording, but as a rule, we avoid future tense. (We can also end sentences with prepositions.)

File: source/upgrade.txt (hunk 3 above)
Comment (mongoKart) [reply source/upgrade.txt:52]: Appreciate the clarification!

File: source/upgrade.txt (hunk 4 above)
Comment (mongoKart) [reply source/upgrade.txt:67]: Debated doing this initially; glad to hear you agree v3.0 information would be more confusing than helpful.

File: source/upgrade.txt (hunk 5 above)
Comment (mongoKart) [reply source/upgrade.txt:76]: We usually try to specify what an entity is, rather than just referring to it by name. What is find_package?

Comment (eramongodb) [reply source/upgrade.txt:78]: [find_package()](https://cmake.org/cmake/help/latest/command/find_package.html) is a CMake command used to import an external CMake package which provides libraries, executables, and other targets as defined by the package. It is the general method for importing external dependencies (that doesn't involve build-from-source via `add_subdirectory`, a la `FetchContent`). It is comparable to `pkg-config` in purpose.

We provide CMake package config files which allow a user to `find_package()` the C++ Driver libraries into their CMake project. The package names are `bsoncxx` and `mongocxx` respectively. Those packages provide the `mongo::bsonxx_*` and `mongo::mongocxx_*` targets corresponding to the provided libraries.

In other words, we dropped support for the legacy CMake package config files (hand-written, variable-based) in favor of modern CMake package config files (generated, target-based).

File: source/upgrade.txt [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+.. _cpp-breaking-changes:
+
+Breaking Changes
+----------------
+
+.. note::
+
+   The {+driver-short+} follows :wikipedia:`semantic versioning. <Software_versioning#Schemes>`
+   The driver might break build system or package compatibility in any release, and
+   patch releases can revert accidental API breaking changes.
```
Comment (eramongodb) [inline source/upgrade.txt:52]:
```suggestion
   The driver may break build system or package compatibility in any release.
   Patch releases may revert accidental API breaking changes.
```
Recommend using "may" instead of "might/can" to imply deliberate policy rather than possibility. Also recommend splitting into individual sentences to better separate the two distinct cases.

File: source/upgrade.txt [hunk 8]
```diff
@@ -0,0 +1,95 @@
...
+
+Version 3.9 Breaking Changes
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+This driver version introduces the following breaking changes:
+
+- Removes support for exported targets from the CMake project build tree.
+- Drops support for macOS 10.14, macOS 10.15, and Ubuntu 14.04.
+- Requires MongoDB C Driver v1.25.0 or later.
+- Requires CMake v3.15 or later
```
Comment (eramongodb) [inline source/upgrade.txt:95]: Rather than committing to a definitive list of breaking changes for 3.8 and earlier or omitting their mention entirely, recommend adding a section like "Version 3.8 and Earlier" that defers to the old release notes on GitHub, e.g.

> ## Version 3.8 and Earlier
>
> For driver versions 3.8 and earlier, see the release notes and associated JIRA tickets for each release: https://github.com/mongodb/mongo-cxx-driver/releases

This way readers will have somewhere to look and be redirected rather than hitting a dead-end.

**Review by eramongodb — APPROVED** [review]
Some minor suggestions; otherwise, LGTM.

File: source/upgrade.txt (hunk 5 above)
Comment (mongoKart) [reply source/upgrade.txt:78]: Makes sense. Thanks!

File: source/upgrade.txt (hunk 8 above)
Comment (mongoKart) [reply source/upgrade.txt:95]: Good idea.

File: source/upgrade.txt (hunk 7 above)
Comment (mongoKart) [reply source/upgrade.txt:52]: Confirmed usage of "may" with my team and this should be fine
